### ISSUE TITLE
RuntimeError (can't modify frozen String):

### ISSUE THREAD

**Issue by allenmackley** [sorentwo/carrierwave-aws#98]
I get "RuntimeError (can't modify frozen String)" when uncommenting this section:

config.aws_signer = -> (unsigned_url, options) do
     Aws::CF::Signer.sign_url(unsigned_url, options)
end

Any idea what I'm doing wrong?

Here's the rest of my config:

config.storage    = :aws
config.aws_bucket = 'my_bucket_name'
config.asset_host = 'https://s3.example.com'
config.aws_authenticated_url_expiration = 60 * 60 * 24 * 7
config.aws_attributes = {
  expires: 1.week.from_now.httpdate,
  cache_control: 'max-age=604800'
}
config.aws_credentials = {
  access_key_id: '... my key id...',
  secret_access_key: '... my secret access key...',
  region: 'us-west-2'
}

Another thing that puzzles me is that my asset host is a CNAME I've configured for CloudFront of https://s3.example.com, however, the URLs that are created end up with the form of https://my_bucket_name.s3-us-west-2.amazonaws.com, which isn't what I've entered in the config...

**Comment by sorentwo** [comment]
There aren't any frozen strings within the carrierwave-aws codebase, so it is probably coming from a dependency. If you have a stacktrace that would really help.

Regarding the asset host issue, it is probably because of permissions. The `asset_host` is only used for public assets (`public-read`).

Pardon the slow reply, just getting back from break now.

**Comment by allenmackley** [comment]
I discovered that CarrierWave doesn't support signed urls for CloudFront. The signed urls are only supported for direct S3 links. I ended up writing sort of a custom workaround and using Fog.

Here's an explanation of what I did on [Stack Exchange](http://stackoverflow.com/questions/9956712/use-cdn-with-carrierwave-fog-in-s3-cloudfront-with-rails-3-1/41353986#41353986). 

I don't know the code for carrierwave-aws but I was just browsing around a little and noticed this. Could this be what's freezing the string?

```
ACCEPTED_ACL = %w[
   private
   public-read
   public-read-write
   authenticated-read
   bucket-owner-read
   bucket-owner-full-control
 ].freeze
```

**Comment by sorentwo** [comment]
That only freezes the array of accepted acl values, it doesn't apply to any of the strings contained within. Even if it did, those are only used to verify the acl value.

**Comment by casiodk** [comment]
I´m getting this error

RuntimeError: can't modify frozen String
from /Users/nicolaiseerup/.rvm/gems/ruby-2.3.1/bundler/gems/cloudfront-signer-a8371aab279f/lib/cloudfront-signer.rb:171:in `gsub!'

here is my config
```

Aws::CF::Signer.configure do |config|
  config.key              = ENV['CDN_PRIVATE_KEY']
  config.key_pair_id      = ENV['CDN_KEY_PAIR_ID']
  config.default_expires  = ENV['CLOUDFRONT_SIGNER_EXPIRES'].to_i
end

CarrierWave.configure do |config|
  config.storage            = :aws
  config.storage            = Rails.env.test? ? :file : :aws
  config.enable_processing  = false if Rails.env.test?

  config.aws_bucket         = "#{ ENV['S3_BUCKET'] }"
  config.aws_acl            = 'public-read'

  # Optionally define an asset host for configurations that are fronted by a
  # content host, such as CloudFront.
  config.asset_host = ENV['CDN_ASSET_HOST'] # if Rails.env.production?

  # The maximum period for authenticated_urls is only 7 days.
  config.aws_authenticated_url_expiration = ENV['AWS_AUTHENTICATED_URL_EXPIRATION'].to_i

  # Set custom options such as cache control to leverage browser caching
  config.aws_attributes = {
    expires: ENV['AWS_ATTRIBUTES_EXPIRES'].to_i.seconds.from_now.httpdate,
    cache_control: "max-age=#{ENV['AWS_ATTRIBUTES_CACHE_CONTROL_MAX_AGE']}"
  }

  config.aws_credentials = {
    access_key_id:     ENV['AWS_ACCESS_KEY_ID'],
    secret_access_key: ENV['AWS_SECRET_ACCESS_KEY'],
    region:            ENV['AWS_REGION']
  }

  # Optional: Signing of download urls, e.g. for serving private
  # content through CloudFront.
  config.aws_signer = -> (unsigned_url, options) { Aws::CF::Signer.sign_url unsigned_url, options } # if Rails.env.production?
end

```

**Comment by sorentwo** [comment]
@casiodk Thanks for the report! Using that info I tracked down the source of the problem and patched it with ff2a911. It turns out that with the `frozen_string_literal` pragma even an interpolated string like what `public_url` generates is automatically frozen.
